---
title: "Storing Data"
output: 
  pdf_document: 
    number_sections: yes
    toc: yes
---

```{r setup, include=FALSE} 
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
```

This dataset can quite easily be handled directly in R, but for larger datasets dimensionality can become a problem. The simplest workaround is to store the data in a SQL query and retrieve only the parts of the data that are needed at any given time.

In here, we will use two files:

* a genotypes file: contains all genotype calls for all SNP and all samples.
* a map file: holds mapping information of the SNP, e.g., chromosome, physical location, etc.

We want to create a database. There are two ways:

# Using command line
For our database we need to create a schema (a schema is simply a text file that describes the database structure and is used to create the initial empty DB). A simple schema will consist of two tables, one for each of the files and one column for each source of information.

```{bash}
#cat Data/snpDB.sql
```

We are ready to create the database, named "SNPsmall":

```{bash}
#sqlite3 Data/SNPsmall < Data/snpDB.sql
```

# Using R
This process does not need an schema.

```{r}
# Load package
library(RSQLite)

# Create a new blank database called SNPsmall
con = dbConnect(dbDriver("SQLite"), dbname = "Output/SNPsmall")

# Upload files to the database
dbWriteTable(
  con,
  "snpmap",
  "Data/SNPmap.txt",
  header = TRUE,
  append = T,
  sep = "\t"
)

dbWriteTable(
  con,
  "SNP",
  "Data/SNPSample.txt",
  append = T,
  header = TRUE,
  skip = 9,
  sep = "\t"
)

# Take a look at the tables and fields in the databse
con = dbConnect(dbDriver("SQLite"), dbname = "Output/SNPsmall")
dbListTables(con)
dbListFields(con, "snpmap")
dbListFields(con, "SNP")
```


You can add indexes to the database to make it faster.
```{r}
dbGetQuery(con, "CREATE INDEX chromosome_idx ON snpmap(chromosome)")
dbGetQuery(con, "CREATE INDEX snp_idx ON SNP(animal)")
dbGetQuery(con, "CREATE INDEX ID_idx ON SNP(snp)")
```

The function **dbGetQuery** is used to send an SQL query to the DB and return the data in a single step.

```{r}
# Retrieve the number of records in a table
dbGetQuery(con, "select count (*) from snpmap")

# Retrieve sample ids
sampleids_s1 = as.vector(dbGetQuery(con, "select distinct animal from SNP"))$animal
head(sampleids_s1)

# Retrieve all data associated with the first sample
hold_s1 = dbGetQuery(con,
                  paste("select * from SNP where animal='", sampleids_s1[1], "'", sep = ""))
head(hold_s1)
```

When we are finished with the database we should close the connection.
```{r}
dbDisconnect(con)
```
