# Introduction
This repo contains my codes on different genomic data analyses. The PDF output of the codes are provided in the *main* branch, while the R Markdown files and the data are provided in the *Notebooks* folder.

# References
Most of the codes and data in this repo come from the following sources:

1- Gondro, C. (2015). Primer to analysis of genomic data using R. New York, NY: Springer.

2- González, J. R., & Cáceres, A. (2019). Omic association studies with r and bioconductor. CRC Press.

3- Irizarry, R. A., & Love, M. I. (2016). Data analysis for the life sciences with R. Chapman and Hall/CRC.

4- PH525x series by HarvardX.

5- Broman, K. W., & Sen, S. (2009). A Guide to QTL Mapping with R/qtl (Vol. 46). New York: Springer.

6- Shi, X. M. (2020). eQTL Analysis. New York: Springer US.
7- Cliff, A., Romero, J., Kainer, D., Walker, A., Furches, A., & Jacobson, D. (2019). A high-performance computing implementation of iterative random forest for the creation of predictive expression networks. Genes, 10(12), 996.
8- Rausch, T., Hsi-Yang Fritz, M., Korbel, J. O., & Benes, V. (2019). Alfred: interactive multi-sample BAM alignment statistics, feature counting and feature annotation for long-and short-read sequencing. Bioinformatics, 35(14), 2489-2491.
9- Erarslan‐Uysal, B., Kunz, J. B., Rausch, T., Richter‐Pechańska, P., van Belzen, I. A., Frismantas, V., ... & Kulozik, A. E. (2020). Chromatin accessibility landscape of pediatric T‐lymphoblastic leukemia and human T‐cell precursors. EMBO molecular medicine, 12(9), e12104.
