---
title: "ATAC-Seq"
output: 
  pdf_document: 
    number_sections: yes
    toc: yes
header-includes: 
  - \usepackage{graphics}
  - \usepackage{float}
editor_options: 
  markdown: 
    wrap: sentence
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(warning = FALSE, message = FALSE)
rm(list = ls())
```

\newpage

# Introduction
Assay for Transposase-Accessible Chromatin with high-throughput sequencing (ATAC-Seq) is a technique for assessing genome-wide chromatin accessibility. The most common use is nucleosome mapping experiments, but it can be applied to mapping transcription factor binding sites, adapted to map DNA methylation sites, or combined with sequencing techniques.

```{r, echo=FALSE, fig.align='center', fig.pos='H', fig.show='hold', out.width = '75%'}
knitr::include_graphics("Figures/16-1.jpeg")
```

# Scenario
We are going to investigate the chromatin accessibility landscape of 13 human primary blood cell types. We will try to identify regulatory regions that govern hematopoietic differentiation. The data is a pre-processed ATAC-Seq count matrix with the raw count matrix of read counts per sample (columns) and peak (rows).

```{bash, echo=FALSE}
# Untar files
tar -zxvf Data/GSE74912_ATACseq_1.txt.tar.gz
tar -zxvf Data/GSE74912_ATACseq_2.txt.tar.gz
tar -zxvf Data/GSE74912_ATACseq_3.txt.tar.gz

# Move files
mv GSE74912_ATACseq_1.txt Output/
mv GSE74912_ATACseq_2.txt Output/
mv GSE74912_ATACseq_3.txt Output/

# Concatenate files
cat Output/GSE74912_ATACseq_1.txt Output/GSE74912_ATACseq_2.txt \
Output/GSE74912_ATACseq_3.txt > Output/GSE74912_ATACseq_All.txt

# Extract blood samples
cat Data/sample.mapping | grep "5483$" | grep -v "CLP" > Output/select.samples
cat Output/select.samples | sed 's/Donor5483-//' | sed 's/-/_/' | sed 's/-[^\t]*//' > Output/blood.samples

# Extract respective columns
COLS=` head -n 1 Output/GSE74912_ATACseq_All.txt | tr '\t' '\n' | grep -n -w -Ff <(cut -f 1 Output/select.samples) | sed 's/:.*//' | tr '\n' ',' | sed 's/,$//'`
cut -f 1-3,${COLS} Output/GSE74912_ATACseq_All.txt | sed 's/Donor5483-//g' | sed 's/-\([^-]*\)-[^\t]*/_\1/g' | gzip -c > Output/atac.data.gz

# Delete extra files
rm Output/GSE74912_ATACseq_1.txt
rm Output/GSE74912_ATACseq_2.txt
rm Output/GSE74912_ATACseq_3.txt
rm Output/GSE74912_ATACseq_All.txt
rm Output/select.samples
```

Let's load the necessary packages first:
```{r}
library(ggplot2)
library(reshape2)
library(pander)
library(Hmisc)
library(pastecs)
library(DESeq2)
library(ChIPseeker)
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
library(clusterProfiler)
library(org.Hs.eg.db)
library(ReactomePA)
library(pheatmap)
```

# Import data
Letâ€™s load the count matrix and the sample information with the blood type and donor:
```{r}
df = read.table("Output/atac.data.gz", skip = 1, header = F, sep = "\t", col.names = c("Chr", "Start", "End", "Bcell-13A", "Bcell-13B", "CD4-9A", "CD4-9B", "CD8-10A", "CD8-10B", "Erythroblast-15A", "Erythroblast-15B", "Erythroblast-15C", "NK-11A", "NK-11B"))
si = read.table("Output/blood.samples", header = F)
colnames(si) = c("sample", "celltype", "donor")
rownames(si) = si$sample
si$donor = factor(si$donor)
pander(dim(df), "Data dimensions")
```

